primitives/state-machine/src/overlayed_changes/mod.rs

```diff
@@ -103,12 +103,35 @@ pub struct OverlayedChanges {
 	children: Map<StorageKey, (OverlayedChangeSet, ChildInfo)>,
 	/// Offchain related changes.
 	offchain: OffchainOverlayedChanges,
+	/// Transaction index changes,
+	transaction_index_ops: Vec<IndexOperation>,
```

Comment:
Just realized while merging master in some PRs, but 'transaction_index_ops' could also support transaction (Same implementation as 'OffchainOverlayedChanges' for offchain indexing).
Without thinking too much about it, I would think it should (would allow conditional indexing).
cc\ @arkpar

Reply:
I did not bother with that for now. As I understand, transactional semantics are used here to support rollback on failing extrinsics. Indexing is transparent to the runtime anyway.  If an extrinsic fails, that fact that it still indexes something does not affect consensus. 
Furthermore:
1. Storage transactions should not be adding to the index until they are pass the failure point. I.e. all fees are applied.
2. Unlike regular transactions, failing storage transactions should not be included in the block in the first place. Storing them for free would defeat the purpose.

Reply:
I don't think its a priority.
I agree consensus is not touched and client can even choose to not index anything at all.
That was the same thing with offchain indexing. Can be implemented at any time.
(I was talking about storage transaction, it is used for different purpose, mainly in contracts afaik)